### ISSUE TITLE
Optimization not progressing with Dask

### ISSUE THREAD

**Issue by AnnaGwen** [AgnostiqHQ/covalent#638]
### Environment

- **Covalent version**:  0.33.0
- **Python version**: 3.8.5
- **Operating system**: MacOS Big Sur 11.6

### What is happening?

Some optimization routines are not actually progressing as expected. In the below script, the cost function should be minimized over the course of 100 iterations, but when it gets dispatched using dask it only returns a constant value.  
The same issue does not occur when it gets dispatched locally. 

### How can we reproduce the issue?

```
import pennylane as qml
from pennylane import qaoa
from pennylane import numpy as np
from matplotlib import pyplot as plt
import networkx as nx
import covalent as ct

from dask.distributed import LocalCluster
cluster=LocalCluster()
dask=ct.executor.DaskExecutor(scheduler_address=cluster.scheduler_address)

@ct.electron
def make_graph(qubits,prob):
    graph = nx.generators.random_graphs.gnp_random_graph(n=qubits,p=prob)
    cost_h, mixer_h = qaoa.min_vertex_cover(graph, constrained=False)
    return cost_h,mixer_h


@ct.electron
def get_circuit(cost_h,mixer_h):
    def qaoa_layer(gamma, alpha):
        qaoa.cost_layer(gamma, cost_h)
        qaoa.mixer_layer(alpha, mixer_h)
        

    def circuit(params,wires, **kwargs):
        depth=params.shape[1]
        for w in range(wires):
            qml.Hadamard(wires=w)
        qml.layer(qaoa_layer, depth, params[0], params[1])
    return circuit

@ct.electron
def make_cost_function(circuit,cost_h,qubits):
    dev = qml.device("lightning.qubit", wires=qubits)
    @qml.qnode(dev)
    def cost_function(params):
        circuit(params,wires=qubits)
        return qml.expval(cost_h)

    return cost_function


@ct.electron
def get_random_initialization(p=1,seed=0):
    np.random.seed(seed)
    return np.random.uniform(0,2*np.pi,(2,p),requires_grad=True)

@ct.electron
def initialize_parameters(p=1,qubits=2,prob=0.3,seed=1):
    cost_h,mixer_h=make_graph(qubits=qubits,prob=prob)
    circuit=get_circuit(cost_h,mixer_h)
    cost_function=make_cost_function(circuit,cost_h,qubits)
    initial_angles=get_random_initialization(p=p,seed=seed)
    return cost_function,initial_angles

@ct.electron
def calculate_cost(cost_function,params,optimizer):
    params,loss=optimizer.step_and_cost(cost_function, params)
    return optimizer,params,loss

@ct.electron
def optimize_electron(cost_function,init_angles, optimizer=qml.GradientDescentOptimizer(),iterations=10):
    loss_history=[]
    params = init_angles
    for _ in range(iterations):
        optimizer,params,loss=calculate_cost(cost_function=cost_function,params=params,optimizer=optimizer)
        loss_history.append(loss)
    return loss_history


@ct.electron
def collect_and_mean(array):
    return np.mean(array,axis=0)

@ct.lattice(executor=dask)
def run_exp(p=1,qubits=2,prob=0.3,seed=[1], optimizers=[qml.GradientDescentOptimizer()],iterations=10):
    compare_optimizers=[]
    for s in seed:
        tmp=[]
        cost_function,init_angles=initialize_parameters(p=p,qubits=qubits,prob=prob,seed=s)
        for optimizer in optimizers:
            loss_history=optimize_electron(cost_function=cost_function,init_angles=init_angles,optimizer=optimizer,iterations=iterations)
            tmp.append(loss_history)
        compare_optimizers.append(tmp)
    return collect_and_mean(compare_optimizers)

id=ct.dispatch(run_exp)(p=1,qubits=2,prob=0.3,seed=[1], optimizers=[qml.GradientDescentOptimizer()],iterations=100)
plt.plot(ct.get_result(id,wait=True).result.T)
plt.show()
```

        
![optimization_issue](https://user-images.githubusercontent.com/10732247/173138250-bbc9a324-0065-4bc7-9ef7-4d49acc9b0bd.png)

### What should happen?

The below image is generated when the same code is ran without using dask.
![optimization_working](https://user-images.githubusercontent.com/10732247/173138572-14fa6a07-502d-4be9-a4a2-aaae67163e81.png)

**Comment by cjao** [comment]
I've also seen this with the Quantum Gravity tutorial (all executors).

**Comment by ruihao-li** [comment]
Yes, like in the MNIST tutorial, I was also training a neural net with PyTorch and I encountered this issue with both the local executor and the Dask executor (see [PR #639](https://github.com/AgnostiqHQ/covalent/pull/639)). I was getting a more or less constant training loss with Covalent. But the training worked fine without using Covalent. 
I will do some more investigations and post any update if I have any.

**Comment by cjao** [comment]
The loss is flat because the [gradients computed](https://github.com/PennyLaneAI/pennylane/blob/e7d52079499508ebf204e4d8678be2fe2f20983e/pennylane/optimize/gradient_descent.py#L59) in `optimizer.step_and_cost()` are empty. Pennylane is using [autograd](https://github.com/PennyLaneAI/pennylane/blob/e7d52079499508ebf204e4d8678be2fe2f20983e/pennylane/_grad.py#L1) to compute the gradient function, and somehow that fails.

A possibly related discussion: https://github.com/dask/distributed/issues/2581#issuecomment-478151764

**Comment by jackbaker1001** [comment]
So @cjao does this work if you choose a derivative free optimizer? (Try Powell or something)

**Comment by cjao** [comment]
The key to this case lies in the following line:

**`cost_function,init_angles=initialize_parameters(p=p,qubits=qubits,prob=prob,seed=s)`**

When the `initialize_parameters` electron is executed using Dask, the returned `init_angles` is merely a `numpy.ndarray` even  `initialize_parameters` constructs the initial angles

```
    initial_angles=get_random_initialization(p=p,seed=seed)
```
as a `pennylane.numpy.tensor.tensor`. Pennylane Tensors have a `requires_grad` attribute whereas ordinary numpy arrays don't, and Pennylane's gradient descent only computes the gradient for [trainable parameters](https://github.com/PennyLaneAI/pennylane/blob/a8feedd65be623237f6da41d7c3a566240165de1/pennylane/_grad.py#L90). 

This problem doesn't occur if the `initialize_parameters` electron uses the local executor, regardless whether  `optimize_electron` uses Dask or local. 

The crux of the issue seems to be that the serialization and deserialization procedures used by Dask to pass data between worker processes somehow converts `pennylane.numpy.tensor.tensor` into a `numpy.ndarray`. The following snippet
```
import pennylane as qml
import pennylane.numpy as np
from dask.distributed import Client, LocalCluster
cluster = LocalCluster()
client = Client(cluster.scheduler_address)

def make_random(p=1):
    return qml.numpy.random.uniform(0,2*np.pi,(2,p),requires_grad=True)

fut = client.submit(make_random)
print(type(fut.result()))
```
returns `<class 'numpy.ndarray'>` for me.

In fact
```
from dask.distributed.protocol import serialize, deserialize

res = qml.numpy.random.uniform(0,2*np.pi,(2,1),requires_grad=True)
res1 = deserialize(*serialize(res))
print(type(res1))
```
yields `<class 'numpy.ndarray'>`.

According to the API docs, [Dask defaults to a custom (de-)serialization protocol but supports others](https://distributed.dask.org/en/stable/serialization.html#defaults), including "msgpack" and "pickle". Choosing "pickle" seems to help:

```
from dask.distributed.protocol import serialize, deserialize

res = qml.numpy.random.uniform(0,2*np.pi,(2,1),requires_grad=True)
res1 = deserialize(*serialize(res, serializers=["pickle"]))
print(type(res1))
```
yields the much more promising `<class 'pennylane.numpy.tensor.tensor'>`.

We can set the protocol on a [per-client basis](https://distributed.dask.org/en/stable/api.html#client).

Setting  `serializers=["pickle"]` [when constructing the Dask client](https://github.com/AgnostiqHQ/covalent/commit/9e8dfecf49c89163d55b80a786509103e7c8e387) seems to resolve this particular issue. 


A possibly related Dask issue: https://github.com/dask/distributed/issues/3716

**Comment by cjao** [comment]
I wonder which of the other training problems (not necessarily using Dask) can also be blamed on serialization/deserialization errors.

**Comment by cjao** [comment]
I wonder which of the other training problems (not necessarily using Dask) can also be traced to serialization/deserialization errors.

**Comment by cjao** [comment]
@santoshkumarradha

**Comment by santoshkumarradha** [comment]
Good catch @cjao , but from what a I remember when I was trying out with PyTorch optimization (a simple optimization) things did work. A temp workaround might be to set the pennylane packets to PyTorch and see @jackbaker1001.

**Comment by cjao** [comment]
The MNIST bug (which uses torch) has a different explanation. Let's discuss in a separate issue.

**Comment by ruihao-li** [comment]
Hey all, I have created another issue related to this when using PyTorch [here](https://github.com/AgnostiqHQ/covalent/issues/645).

**Comment by cjao** [comment]
Until #648 is addressed, a temporary fix is to use `executor=local` for the `initialize_parameters` electron.

**Comment by cjao** [comment]
Potentailly addreseed in [#748](https://github.com/AgnostiqHQ/covalent/issues/748#issuecomment-1172934487)
